torch/distributions/transforms.py

```diff
@@ -652,9 +657,9 @@ def log_abs_det_jacobian(self, x, y):
         for trans, length in zip(self.transforms, self.lengths):
             xslice = x.narrow(self.dim, start, length)
             yslice = y.narrow(self.dim, start, length)
-            logdetjacs.append(trans.log_abs_det_jacobian(xslice, yslice))
+            logdetjacs.append(trans.log_abs_det_jacobian(xslice, yslice).reshape(x.shape + (1,) * self.event_dim))
             start = start + length  # avoid += for jit compat
-        return torch.cat(logdetjacs, dim=self.dim)
+        return _sum_rightmost(torch.cat(logdetjacs, dim=self.dim), self.event_dim)
```

Comment:
Hmm I'm worried this has the order of operations incorrect. I would think you would want
```py
for trans, length in zip(self.transforms, self.lengths):
    xslice = x.narrow(self.dim, start, length)
    yslice = y.narrow(self.dim, start, length)
    logdetjac = trans.log_abs_det_jacobian(xslice, yslice)
    if trans.event_dim < self.event_dim:
        logdetjac = _sum_rightmost(logdetjac, self.event_dim - trans.event_dim)
    logdetjacs.append(logdetjac)
return torch.cat(logdetjac, dim=self.dim)
```

Reply:
I think the issue here is that the join dim for `CatTransform` can be one of the dimensions that `log_abs_det_jacobian` has summed out when `event_dim` > 0.

So my logic was: add the dimensions that `log_abs_det_jacobian` sums out, do the concatenation which is now guaranteed to succeed, then sum out the dimensions again to give the correct output shape (which will be different for this function than `forward` and `inverse` when `event_dim` > 0).

Won't doing the concatenation last fail since `self.dim > len(logdetjac)` can be true?

Reply:
Hmm I think I understand, but your proposal of expanding would still overcount. I think we would need to branch on whether the concatenated dim is an event dim or a batch dim. If an event dim, we sum; if an event dim we concat the log abs det jacobians:
```py
for trans, length in zip(self.transforms, self.lengths):
    xslice = x.narrow(self.dim, start, length)
    yslice = y.narrow(self.dim, start, length)
    logdetjac = trans.log_abs_det_jacobian(xslice, yslice)
    if trans.event_dim < self.event_dim:
        logdetjac = _sum_rightmost(logdetjac, self.event_dim - trans.event_dim)
    logdetjacs.append(logdetjac)
# Decide whether to concatenate or sum.
dim = self.dim
if dim >= 0:
    dim = dim - x.dim()
dim = dim + self.event_dim
if dim < 0:
    return torch.cat(logdetjac, dim=dim)
else:
    return sum(logdetjac)
```

Reply:
For a simple example to see double counting of the original proposal, consider
```py
two = torch.ones(2) * 2
CatTransform([AffineTransform(0, two, event_dim=0),
              AffineTransform(0, two, event_dim=1)])
```
Now the correct log abs det jacobian should be `4 * log(2)`, but in the original implementation provided, the transform with `event_dim=1` is combined with `_sum_rightmost`, then expanded by `(1,)`s, then broadcasted to again have shape `(2,)`, then summed again with `_sum_rightmost`, leading to final result of `6 * log(2)`.

@stefanwebb why don't we add this example as a unit test?

Reply:
I think I need to take some more time to fix this, make sure it passes tests, etc. I'll ping you when it's ready for a second review!

Reply:
One thing I didn't consider is when the transforms have different `event_dim`'s, like you point out, and it would need a term like `self.event_dim - trans.event_dim`